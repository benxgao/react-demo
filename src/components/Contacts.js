import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Contact from './Contact';
import Preloader from './Preloader';

const Contacts = ({
  activedContact,
  contacts,
  isLoading,
  onContactDetails,
}) => {
  const onClick = (contact) => {
    onContactDetails(contact);
  };

  const preloaderView = (
    <div className="preloader-container">
      <Preloader />
    </div>
  );

  const isContactShowing = contact => contact.get('id') === activedContact.get('id');

  const listItemStyles = contact => classnames({
    contacts__li: true,
    'contacts__li--active': isContactShowing(contact),
  });

  const nameCardPanelStyles = contact => classnames({
    'contacts__namecard-panel': true,
    'contacts__namecard-panel--active': isContactShowing(contact),
  });

  const contactList = contacts.map((contact) => {
    const id = contact.get('id');
    const name = contact.get('name');
    const classes = nameCardPanelStyles(contact);

    return (
      <div key={id}>
        <div className={listItemStyles(contact)} onClick={() => onClick(contact)}>
          <div className="contacts__li-name">{name}</div>
        </div>
        <Contact contact={contact} classes={classes} />
      </div>
    );
  });

  const contactsView = (
    <div className="contacts">
      {contactList}
    </div>
  );

  return (
    <div className="container">
      <div className="row">
        {isLoading ? preloaderView : contactsView}
      </div>
    </div>
  );
};

Contacts.propTypes = {
  activedContact: PropTypes.object.isRequired,
  contacts: PropTypes.any.isRequired,
  isLoading: PropTypes.bool.isRequired,
  onContactDetails: PropTypes.func.isRequired,
};

Contacts.displayName = 'Contacts';

export default Contacts;
