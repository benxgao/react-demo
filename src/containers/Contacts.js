import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import Contacts from '../components/Contacts';
import { onContactDetails } from '../actions/AppAction';

const mapStateToProps = state => ({
  activedContact: state.app.get('activedContact'),
  contacts: state.app.get('contacts'),
  isLoading: state.app.get('isLoading'),
});

const mapDispatchToProps = dispatch => bindActionCreators({
  onContactDetails,
}, dispatch);

const mergeProps = (stateProps, dispatchProps, ownProps) => Object.assign(
  {}, ownProps, stateProps, dispatchProps
);

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)(Contacts);
