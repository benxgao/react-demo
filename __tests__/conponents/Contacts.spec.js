import Immutable from 'immutable';
import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
// import { mount } from 'enzyme';
// import { mountToJson } from 'enzyme-to-json';
import Contacts from '../../src/components/Contacts';
import dataMock from '../../__mocks__/dataMock';

const contactStub = dataMock[0];
const propsStub = {
  activedContact: Immutable.fromJS(contactStub),
  contacts: Immutable.fromJS([contactStub]),
  isLoading: false,
  onContactDetails: jest.fn(),
};

describe('<Contacts />', () => {
  describe('render()', () => {
    it('should render <Preloader /> when isLoading is true', () => {
      const props = Object.assign({}, propsStub, {
        isLoading: true,
      });
      const wrapper = shallow(
        <Contacts {...props} />
      );

      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('should render <Contact /> list when isLoading is false', () => {
      const props = Object.assign({}, propsStub);
      const wrapper = shallow(
        <Contacts {...props} />
      );

      expect(shallowToJson(wrapper.find('.contacts'))).toMatchSnapshot();
    });
  });

  describe('onClick()', function() {
    it('should call onContactDetails function', () => {
      const spy = jest.fn();
      const props = Object.assign({}, propsStub, {
        onContactDetails: spy
      });
      const wrapper = shallow(
        <Contacts {...props} />
      );

      wrapper.find('.contacts__li').simulate('click');
      expect(spy).toHaveBeenCalledWith(Immutable.fromJS(contactStub));
    });
  });
});
