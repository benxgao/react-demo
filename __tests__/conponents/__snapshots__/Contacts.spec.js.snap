// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Contacts /> render() should render Contact list when isLoading is false 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="contacts"
    >
      <div>
        <div
          className="contacts__li contacts__li--active"
          onClick={[Function]}
        >
          <div
            className="contacts__li-name"
          >
            Leanne Graham
          </div>
        </div>
        <Contact
          classes="contacts__namecard-panel contacts__namecard-panel--active"
          contact={
            Object {
              "address": Object {
                "city": "Gwenborough",
                "geo": Object {
                  "lat": "-37.3159",
                  "lng": "81.1496",
                },
                "street": "Kulas Light",
                "suite": "Apt. 556",
                "zipcode": "92998-3874",
              },
              "company": Object {
                "bs": "harness real-time e-markets",
                "catchPhrase": "Multi-layered client-server neural-net",
                "name": "Romaguera-Crona",
              },
              "email": "Sincere@april.biz",
              "id": 1,
              "name": "Leanne Graham",
              "phone": "1-770-736-8031 x56442",
              "username": "Bret",
              "website": "hildegard.org",
            }
          }
        />
      </div>
      <div>
        <div
          className="contacts__li"
          onClick={[Function]}
        >
          <div
            className="contacts__li-name"
          >
            Ervin Howell
          </div>
        </div>
        <Contact
          classes="contacts__namecard-panel"
          contact={
            Object {
              "address": Object {
                "city": "Wisokyburgh",
                "geo": Object {
                  "lat": "-43.9509",
                  "lng": "-34.4618",
                },
                "street": "Victor Plains",
                "suite": "Suite 879",
                "zipcode": "90566-7771",
              },
              "company": Object {
                "bs": "synergize scalable supply-chains",
                "catchPhrase": "Proactive didactic contingency",
                "name": "Deckow-Crist",
              },
              "email": "Shanna@melissa.tv",
              "id": 2,
              "name": "Ervin Howell",
              "phone": "010-692-6593 x09125",
              "username": "Antonette",
              "website": "anastasia.net",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<Contacts /> render() should render Contact list when isLoading is true 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="preloader-container"
    >
      <Preloader />
    </div>
  </div>
</div>
`;
